### PR TITLE
Fixed memory corruption with UTM coordinates search

### LINKED ISSUE #7227: Crash with memory tagging (MTE) enabled
**Describe the issue**
I was opening Organic Maps, when it crashed at startup. It only happened once, and I didn't manage to reproduce it.

**Steps to reproduce**
1. Get a Pixel8 phone with MTE enabled
2. Open Organic Maps
3. Maybe get a crash

**Expected behaviour**
No crash

**System information:**
 - Operating system and its version: [GrapheneOS 2024011600](https://grapheneos.org/releases#2024011600)
 - Organic Maps version: 2024.01.09-6-Google
 - Device Model: Pixel8

**Additional context**

```
type: crash
osVersion: google/shiba/shiba:14/UQ1A.240105.004/2024011600:user/release-keys
package: app.organicmaps:24010906
process: app.organicmaps
processUptime: 0 + 0 ms
installer: com.android.vending

signal 11 (SIGSEGV), code 9 (SEGV_MTESERR), fault addr 0x0300d4ec8127c8bf

backtrace:
      #00 pc 0000000000d98128  /data/app/~~I_0G_hPjxDo4_xS0_8naZw==/app.organicmaps-nLh_dYfOtY6_PnahPJN0gA==/split_config.arm64_v8a.apk!liborganicmaps.so (offset 0x1000) (search::MatchUTMCoords(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&)+352) (BuildId: 6737418c75baa9a336c6840dfaa1ffa4f2f29d58)
      #01 pc 0000000000e1b8b8  /data/app/~~I_0G_hPjxDo4_xS0_8naZw==/app.organicmaps-nLh_dYfOtY6_PnahPJN0gA==/split_config.arm64_v8a.apk!liborganicmaps.so (offset 0x1000) (search::Processor::SearchCoordinates()+128) (BuildId: 6737418c75baa9a336c6840dfaa1ffa4f2f29d58)
      #02 pc 0000000000dde218  /data/app/~~I_0G_hPjxDo4_xS0_8naZw==/app.organicmaps-nLh_dYfOtY6_PnahPJN0gA==/split_config.arm64_v8a.apk!liborganicmaps.so (offset 0x1000) (search::Processor::Search(search::SearchParams)+1576) (BuildId: 6737418c75baa9a336c6840dfaa1ffa4f2f29d58)
      #03 pc 0000000000ddd758  /data/app/~~I_0G_hPjxDo4_xS0_8naZw==/app.organicmaps-nLh_dYfOtY6_PnahPJN0gA==/split_config.arm64_v8a.apk!liborganicmaps.so (offset 0x1000) (search::Engine::DoSearch(search::SearchParams, std::__ndk1::shared_ptr<search::ProcessorHandle>, search::Processor&)+540) (BuildId: 6737418c75baa9a336c6840dfaa1ffa4f2f29d58)
      #04 pc 0000000000e4bb30  /data/app/~~I_0G_hPjxDo4_xS0_8naZw==/app.organicmaps-nLh_dYfOtY6_PnahPJN0gA==/split_config.arm64_v8a.apk!liborganicmaps.so (offset 0x1000) (BuildId: 6737418c75baa9a336c6840dfaa1ffa4f2f29d58)
      #05 pc 0000000000ddaa1c  /data/app/~~I_0G_hPjxDo4_xS0_8naZw==/app.organicmaps-nLh_dYfOtY6_PnahPJN0gA==/split_config.arm64_v8a.apk!liborganicmaps.so (offset 0x1000) (search::Engine::MainLoop(search::Engine::Context&)+268) (BuildId: 6737418c75baa9a336c6840dfaa1ffa4f2f29d58)
      #06 pc 00000000010f7318  /data/app/~~I_0G_hPjxDo4_xS0_8naZw==/app.organicmaps-nLh_dYfOtY6_PnahPJN0gA==/split_config.arm64_v8a.apk!liborganicmaps.so (offset 0x1000) (threads::SimpleThread::ThreadFunc(std::__ndk1::function<void ()>&&)+36) (BuildId: 6737418c75baa9a336c6840dfaa1ffa4f2f29d58)
      #07 pc 0000000000e4addc  /data/app/~~I_0G_hPjxDo4_xS0_8naZw==/app.organicmaps-nLh_dYfOtY6_PnahPJN0gA==/split_config.arm64_v8a.apk!liborganicmaps.so (offset 0x1000) (BuildId: 6737418c75baa9a336c6840dfaa1ffa4f2f29d58)
      #08 pc 00000000000d006c  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+204) (BuildId: 3bcd4bc92892dc0d588978c3b951b7f8)
      #09 pc 0000000000064db0  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64) (BuildId: 3bcd4bc92892dc0d588978c3b951b7f8)
```

### PR DESCRIPTION
**Pull request by biodranik**
Fixes #7227

### CODE REVIEW

File: search/utm_mgrs_coords_match.cpp [hunk 1]
```diff
@@ -4,74 +4,67 @@
...
+static std::string_view constexpr kSpaceChars = " \t\r\n";
 
 // Matches 2 digits zone code. Returns end position of matched chars or string::npos if no match.
-size_t MatchZoneCode(string const & query, int & zone_code)
+static size_t MatchZoneCode(std::string const & query, int & zoneCode)
 {
   auto const pos = query.find_first_not_of(kSpaceChars);
-  if (query.size() - pos < 2)
-    return string::npos;
+  if (pos == std::string::npos || query.size() < pos + 2)
```
Comment (biodranik) [inline search/utm_mgrs_coords_match.cpp:20]: Random memory access

File: search/utm_mgrs_coords_match.cpp [hunk 2]
```diff
@@ -4,74 +4,67 @@
...
 
 
 // Matches zone letter ignoring spaces. Returns end position of matched chars or string::npos if no match.
-size_t MatchZoneLetter(string const & query, char & zone_letter, size_t startPos)
+size_t MatchZoneLetter(std::string const & query, char & zoneLetter, size_t startPos)
 {
   auto const pos = query.find_first_not_of(kSpaceChars, startPos);
-  if (query.size() == pos)
-    return string::npos;
+  if (pos == std::string::npos || query.size() == pos)
```
Comment (biodranik) [inline search/utm_mgrs_coords_match.cpp:37]: Random memory access

File: search/utm_mgrs_coords_match.cpp [hunk 3]
```diff
@@ -4,74 +4,67 @@
...
 }
 
 // Matches long number ignoring spaces. Returns end position of matched chars or string::npos if no match.
-size_t MatchInt(string const & query, int & value, size_t startPos)
+size_t MatchInt(std::string const & query, int & value, size_t startPos)
 {
   auto pos = query.find_first_not_of(kSpaceChars, startPos);
-  if (query.size() == pos)
-    return string::npos;
+  if (pos == std::string::npos || query.size() == pos)
```
Comment (biodranik) [inline search/utm_mgrs_coords_match.cpp:52]: Random memory access

File: platform/utm_mgrs_utils.cpp [hunk 4]
```diff
@@ -209,52 +209,48 @@ string get_100k_id(double easting, double northing, int zone_number)
...
+  return strings::to_string_width(point.zoneNumber, 2) + point.zoneLetter + " " +
+         Get100kId(point.easting, point.northing, point.zoneNumber) + " " +
+         eastingStr.substr(1, precision) + " " +
+         northingStr.substr(1, precision);
 }
+}  // namespace
 
 // Convert UTM parameters to lat,lon for WSG 84 ellipsoid.
 // If UTM parameters are valid lat and lon references are used to output calculated coordinates.
 // Otherwise function returns 'false'.
```
Comment (strump) [inline platform/utm_mgrs_utils.cpp:236]:
```suggestion
// Otherwise function returns empty optional.
```

File: platform/utm_mgrs_utils.cpp [hunk 5]
```diff
@@ -209,52 +209,48 @@ string get_100k_id(double easting, double northing, int zone_number)
...
+    northingStr = northingStr.substr(northingStr.size() - 6);
 
-    return strings::to_string_width((long) point.zone_number, 2) + point.zone_letter + " " + \
-           get_100k_id(point.easting, point.northing, point.zone_number) + " " + \
-           eastingStr.substr(1, precision) + " " + \
-           northingStr.substr(1, precision);
-  }
+  return strings::to_string_width(point.zoneNumber, 2) + point.zoneLetter + " " +
+         Get100kId(point.easting, point.northing, point.zoneNumber) + " " +
+         eastingStr.substr(1, precision) + " " +
```
Comment (strump) [inline platform/utm_mgrs_utils.cpp:229]: Should we replace `" "` with `' '` as you did at lines 153-155.

Comment (biodranik) [inline platform/utm_mgrs_utils.cpp:229]:
```suggestion
  return strings::to_string_width(point.zoneNumber, 2) + point.zoneLetter + ' ' +
         Get100kId(point.easting, point.northing, point.zoneNumber) + ' ' +
         eastingStr.substr(1, precision) + ' ' +
```

**Review by strump — APPROVED** [review]

**Pull request merged**